RecoTracker/DeDx/plugins/DeDxHitCalibrator.cc

```diff
@@ -0,0 +1,427 @@
+#include "CondFormats/DataRecord/interface/DeDxCalibrationRcd.h"
+#include "CondFormats/SiPixelObjects/interface/PixelIndices.h"
+
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "Geometry/Records/interface/TrackerTopologyRcd.h"
+#include "Geometry/CommonDetUnit/interface/PixelGeomDetUnit.h"
+#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+#include "Geometry/TrackerGeometryBuilder/interface/StripGeomDetUnit.h"
+#include "CalibTracker/SiPixelESProducers/interface/SiPixelGainCalibrationOfflineService.h"
+
+#include <fstream>
+
+class DeDxHitCalibrator : public edm::stream::EDProducer<> {
+public:
+  static constexpr int kIsNormal = 0, kIsBelow = 1, kIsOver = 2;
+  static constexpr int PXB = 0, PXF = 1, TIB = 2, TID = 3, TOB = 4, TECThin = 5, TECThick = 6;
+  typedef std::pair<uint32_t, unsigned char> ChipId;
+
+  explicit DeDxHitCalibrator(const edm::ParameterSet&);
+  ~DeDxHitCalibrator() override = default;
+  static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+private:
+  void beginRun(edm::Run const&, const edm::EventSetup&) override;
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+  int getDetId(const DetId&, const float&);
+  float correctEnergy(const float&, const ChipId&);
+  void processHitInfo(const reco::DeDxHitInfo&,
+                      const float& trackMomentum,
+                      reco::DeDxHitCollection&,
+                      reco::DeDxHitCollection&);
+
+  double getChi2(const std::vector<double>&, const std::vector<std::pair<double, int> >&, const double&, const double&);
+  void getAlphaBeta(const std::vector<double>&,
+                    const std::vector<std::pair<double, int> >&,
+                    CLHEP::HepMatrix&,
+                    CLHEP::HepVector&,
+                    const std::vector<bool>&,
+                    const double&,
+                    const double&);
+  std::pair<double, double> fitStripCluster(const std::vector<std::pair<double, int> >&, const double&, const double&);
+
+  const bool applyGain_;
+  const double MeVPerElectron_;
+  const int VCaltoElectronGain_, VCaltoElectronGain_L1_, VCaltoElectronOffset_, VCaltoElectronOffset_L1_;
+  const int pixelSaturationThr_;
+  const edm::EDGetTokenT<reco::TrackCollection> tracksToken_;
+  const edm::EDGetTokenT<reco::DeDxHitInfoAss> dedxHitInfoToken_;
+  const edm::ESGetToken<DeDxCalibration, DeDxCalibrationRcd> dedxCalibToken_;
+  const edm::ESGetToken<TrackerGeometry, TrackerDigiGeometryRecord> tkGeomToken_;
+  const edm::ESGetToken<TrackerTopology, TrackerTopologyRcd> tkTopoToken_;
+
+  SiPixelGainCalibrationOfflineService pixelCalib_;
+  edm::ESHandle<DeDxCalibration> dedxCalib_;
+  edm::ESHandle<TrackerGeometry> tkGeom_;
+  edm::ESHandle<TrackerTopology> tkTopo_;
+
+  // Energy loss parametrization from https://doi.org/10.1016/j.nima.2012.06.064
+  static constexpr double a_ = 0.07;     //energy loss factor in silicon
```

Comment:
Why are these two `static constexpr` defined here rather than in `processHitInfo`, which is the only place where they are used?

Reply:
No specific reason. I will move them down to where they are used

Reply:
The constexpr are moved to processHitInfo in 2ed86ee6a3d96c4c7bd1fa579f3089e0bed8895e